drivers/gpio/gpio_mcp23xxx.c

```diff
@@ -0,0 +1,296 @@
+		LOG_ERR("Error setting pin direction (%d)", ret);
+		goto done;
+	}
+
+	ret = setup_pin_pull(dev, pin, flags);
+	if (ret < 0) {
+		LOG_ERR("Error setting pin pull up/pull down (%d)", ret);
+		goto done;
+	}
+
+done:
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp23xxx_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	struct mcp23xxx_drv_data *drv_data = (struct mcp23xxx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	ret = read_port_regs(dev, REG_GPIO, &buf);
+	if (ret == 0) {
+		*value = buf;
+	}
+
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp23xxx_port_set_masked_raw(const struct device *dev, uint32_t mask, uint32_t value)
+{
+	struct mcp23xxx_drv_data *drv_data = (struct mcp23xxx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	buf = drv_data->reg_cache.gpio;
+	buf = (buf & ~mask) | (mask & value);
+
+	ret = write_port_regs(dev, REG_GPIO, buf);
+	if (ret == 0) {
+		drv_data->reg_cache.gpio = buf;
+	}
+
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp23xxx_port_set_bits_raw(const struct device *dev, uint32_t mask)
+{
+	return mcp23xxx_port_set_masked_raw(dev, mask, mask);
+}
+
+static int mcp23xxx_port_clear_bits_raw(const struct device *dev, uint32_t mask)
+{
+	return mcp23xxx_port_set_masked_raw(dev, mask, 0);
+}
+
+static int mcp23xxx_port_toggle_bits(const struct device *dev, uint32_t mask)
+{
+	struct mcp23xxx_drv_data *drv_data = (struct mcp23xxx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	buf = drv_data->reg_cache.gpio;
+	buf ^= mask;
+
+	ret = write_port_regs(dev, REG_GPIO, buf);
+	if (ret == 0) {
+		drv_data->reg_cache.gpio = buf;
+	}
+
+	k_sem_give(&drv_data->lock);
+
+	return ret;
+}
+
+static int mcp23xxx_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+					    enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	return -ENOTSUP;
+}
+
+const struct gpio_driver_api gpio_mcp23xxx_api_table = {
+	.pin_configure = mcp23xxx_pin_cfg,
+	.port_get_raw = mcp23xxx_port_get_raw,
+	.port_set_masked_raw = mcp23xxx_port_set_masked_raw,
+	.port_set_bits_raw = mcp23xxx_port_set_bits_raw,
+	.port_clear_bits_raw = mcp23xxx_port_clear_bits_raw,
+	.port_toggle_bits = mcp23xxx_port_toggle_bits,
+	.pin_interrupt_configure = mcp23xxx_pin_interrupt_configure,
+};
+
+/**
+ * @brief Initialization function of MCP23XXX
+ *
+ * @param dev Device struct.
+ * @return 0 if successful. Otherwise <0 is returned.
+ */
+int gpio_mcp23xxx_init(const struct device *dev)
+{
+	const struct mcp23xxx_config *config = dev->config;
+	struct mcp23xxx_drv_data *drv_data = (struct mcp23xxx_drv_data *)dev->data;
+	int err;
+
+	if (config->ngpios != 8 && config->ngpios != 16) {
```

Comment:
nit: add `U` suffix to all unsigned constants

```suggestion
	if (config->ngpios != 8U && config->ngpios != 16U) {
```

Reply:
Did that for all the 8 and 16 unsigned constants, there were a few comparisons to `0` elsewhere that was code copied from existing places, not sure if those should be updated or not.